using HtmlGenerator;

namespace HTMLNameSpace
{
    /// <summary>
    /// To contain all help messages related to tables
    /// </summary>
    public class HTMLHelp
    {
        public static HtmlBuilder AnnotatedConsensusSequence = new HtmlBuilder(HtmlTag.p, "The consensus sequence aligned with the germline sequence. The differences are accentuated by displaying the changes from the germline sequence below. The format follows the format of Domain Gap Align.");
        public static HtmlBuilder ConsensusSequence = new HtmlBuilder(HtmlTag.p, "The sequence obtained by taking the highest scoring amino acid for each position. The score is calculated as the sum of all positional scores for that amino acid on this position. Any position where no reads where mapped are filled in by the sequence from the template, otherwise the template does not have any direct effect on the scoring. Insertions into the consensus sequence are only taken into account if the total score for that insertion is higher than the 'default' score, which is calculated by summing all positional scores of all amino acids not followed by an insertion from the previous position.");
        public static HtmlBuilder DOCGraph = new HtmlBuilder(HtmlTag.p, "The depth of coverage reported is the sum of all positional scores for that position. This is reported with the same position numbering as the consensus sequence.");
        public static HtmlBuilder Order = new HtmlBuilder(HtmlTag.p, "The chosen templates for this recombination.");
        public static HtmlBuilder OverviewOfScores = new HtmlBuilder(HtmlTag.p, "The scores for all templates in this segment. All groups are displayed as a transparent box with its name in the right hand corner. The groups are sorted on their highest score. Within each group they are sorted on the score. You can use the buttons to toggle certain scores and their unique counterparts on or off.");
        public static HtmlBuilder ReadLookup = new HtmlBuilder(HtmlTag.p, "All places where this read could be placed.");
        public static HtmlBuilder ReadsAlignment = new HtmlBuilder(HtmlTag.p, "The exported data is a FASTA file containing all reads spaced by '~' and gaps are indicated by '.'. The first sequence is the template and has as header '>{id} template'. All other sequences are reads which have as header '>{id} score:{score} alignment:{alignment/CIGAR}'.");
        public static HtmlBuilder RecombinedSequence = new HtmlBuilder(HtmlTag.p, "The recombined sequence as generated from the Templates that can be seen under 'Order'.");
        public static HtmlBuilder SegmentJoining = new HtmlBuilder(HtmlTag.p, "This are the scores for all tested overlaps. The X-axis displays the overlaps the Y-axis the scores. If there are multiple overlaps with a high score make sure to manually check if stitch did indeed pick the best one.");
        public static HtmlBuilder SequenceConsensusOverview = new HtmlBuilder(HtmlTag.p, "This displays the found diversity of amino acids for each position. The amino acids are linearly scaled to the respective score. The score is calculated as the sum of all positional scores (if present otherwise the general scores) for this position in all aligned reads. Any positions where no reads where mapped are displayed as dots '.'.");
        public static HtmlBuilder SequenceConsensusOverviewData = new HtmlBuilder(HtmlTag.p, "A TSV file with for each position in the consensus sequence all found amino acids with its score. This par is saved as two consecutive columns filled with first the amino acid and second the score.");
        public static HtmlBuilder Spectrum = new HtmlBuilder(HtmlTag.p, "The raw spectrum of this peptide. The fragments are coloured according to ion type (see legend). Any peaks with a star '*' as text can be hovered over to see the full details, first the ion type second the mass shift type. By hovering over the amino acids in the peptide or ions in the legend the corresponding peaks are highlighted. By toggling the 'Unassigned' label you can turn the background (unassigned) peaks on or off in the plot. By updating the slider in the Ion legend you can update the spectrum to only show the top X% of the peaks with labels. The top X% means any peak that is within X% of the highest intensity. By dragging in the spectrum you can zoom in to a specific part of the spectrum and use 'Zoom Out' to get back to the original zoom level.");
        public static HtmlBuilder TemplateIdentifier = new HtmlBuilder(HtmlTag.p, "The identifier for this template.");
        public static HtmlBuilder TemplateLength = new HtmlBuilder(HtmlTag.p, "The length of the sequence in amino acids, excluding any 'X's added by 'GapHead' or 'GapTail'.");
        public static HtmlBuilder TemplateMatches = new HtmlBuilder(HtmlTag.p, "The total number of placed reads on this template.");
        public static HtmlBuilder TemplateScore = new HtmlBuilder(HtmlTag.p, "The total score calculated by summing all sums for all placed reads on this template.");
        public static HtmlBuilder TemplateSequence = new HtmlBuilder(HtmlTag.p, "The template sequence as read from the original file (without any annotations).");
        public static HtmlBuilder TemplateTotalArea = new HtmlBuilder(HtmlTag.p, "The sum of the area of all amino acids placed on this template.");
        public static HtmlBuilder TemplateUniqueArea = new HtmlBuilder(HtmlTag.p, "The sum of the area of all amino acids uniquely placed on this templates.");
        public static HtmlBuilder TemplateUniqueMatches = new HtmlBuilder(HtmlTag.p, "The total number of reads uniquely placed on this template.");
        public static HtmlBuilder TemplateUniqueScore = new HtmlBuilder(HtmlTag.p, "The total score calculated by summing all sums for all reads uniquely placed on this template.");
        public static HtmlBuilder Tree = new HtmlBuilder(HtmlTag.p, "This tree is made on the fly based on the original sequences of all templates in this segment. The tree is made using the Neighbour Joining algorithm and rooted using a generated random protein sequence. To find the score for each node all reads are placed on the tree at the highest node where they are placed at every child template. If a read is only placed on a single template (placed uniquely) it is shown in the left half of the circle on that template. You can change the displayed score to the total matches or total area using the buttons above the tree.");
        public static HtmlBuilder TreeData = new HtmlBuilder(HtmlTag.p, "The tree is saved in JSON format with an object for each node with 'left' and 'right' members for the corresponding branches in the tree. The nodes also contain the calculated distances for both branches and the score, matches, and area for each node. The leaf nodes contain all scores including their unique counterparts and the name of the leaf.");
    }
}